### PR TITLE
Transformers: fix src and key padding mask bool regression

### LINKED ISSUE #95702: _canonical_mask throws warning when bool masks passed as input to TransformerEncoder
### 🐛 Describe the bug

Passing bool masks to `TransformerEncoder` causes warnings to be raised:

```
import torch
import torch.nn as nn

encoder_layer = nn.TransformerEncoderLayer(
    d_model=16,
    nhead=2,
    dim_feedforward=32,
    dropout=0.1,
    activation='relu',
    batch_first=True,
)
encoder_norm = nn.LayerNorm(16)
encoder = nn.TransformerEncoder(
    encoder_layer, 2, encoder_norm
)

inputs = torch.randn(2,3,16)

src_mask = torch.ones(3, 3, dtype=torch.bool).triu_(diagonal=1)
input_seq_len = torch.tensor([3,2])
padding_mask = (torch.arange(3)[None, :].cpu() >= input_seq_len[:, None])

assert(src_mask.dtype == padding_mask.dtype)

encoder(inputs, 
    mask=src_mask,
    src_key_padding_mask=padding_mask,
)
```

> ~/research/pt2/lib/python3.9/site-packages/torch/nn/functional.py:5004: UserWarning: Support for mismatched src_key_padding_mask and src_mask is deprecated. Use same type for both instead.
> ~/research/pt2/lib/python3.9/site-packages/torch/nn/functional.py:5004: UserWarning: Support for mismatched key_padding_mask and attn_mask is deprecated. Use same type for both instead.

These warnings should not be raised, since the two masks are the same dtype.

It looks like `F._canonical_mask` gets called 3 times, by `TransformerEncoder`, `TransformerEncoderLayer` and `MultiHeadAttention`: the first call converts a bool padding mask to float, which then doesn't match the src_mask on subsequent calls.

### Versions

Collecting environment information...                                                                                                                                        
PyTorch version: 2.0.0.dev20230228+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.9.13 (main, Oct 13 2022, 21:15:33)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-60-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 525.78.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          8
On-line CPU(s) list:             0-7
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz
CPU family:                      6
Model:                           158
Thread(s) per core:              1
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        13
CPU max MHz:                     4700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_
perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_
timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2
 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities
Virtualisation:                  VT-x
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        2 MiB (8 instances)
L3 cache:                        12 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT disabled                                                                                                 
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Mitigation; Microcode
Vulnerability Tsx async abort:   Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.0.0+b8b470bc59
[pip3] torch==2.0.0.dev20230228+cu118
[pip3] torchaudio==2.0.0.dev20230223+cu118
[pip3] torchvision==0.15.0.dev20230227+cu118
[conda] numpy                     1.21.5                   pypi_0    pypi
[conda] torch                     1.7.1                    pypi_0    pypi
[conda] torchseq                  3.0.0a0                   dev_0    <develop>

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @albanD @mruberry @walterddr @saketh-are

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: fix src and pad mask bool regression

This fixes a regression introduced previously with #92733. That PR unified testing of masks to remove Byte Tensors as permissible mask, introduced mask compatibility check, and mask conversion to FP mask.  The problem addressed in this PR was that after the first mask had been converted, a check for mask compatibility would fail.  

Test Plan: sandcastle & github

Differential Revision: D43782858

Fixes  https://github.com/pytorch/pytorch/issues/95702

### CODE REVIEW

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:2114]: can you move this test to the TestTransformers class

**Comment by mikekgfb** [comment]
Fix issue 95702

File: test/test_transformers.py
Comment (malfet) [inline test/test_transformers.py:2146]: Nit
```suggestion
```

File: torch/nn/functional.py
Comment (malfet) [inline torch/nn/functional.py:5210]: Is there a reason to change `other_name`? (I guess it would affect serialization, wouldn't it?)

File: test/test_transformers.py
Comment (cpuhrsch) [inline test/test_transformers.py:538]: What is this meant to test?

File: test/test_transformers.py
Comment (cpuhrsch) [inline test/test_transformers.py:536]:
```suggestion
        self.assertEqual(src_mask.dtype, padding_mask.dtype)
```

**Review by cpuhrsch — APPROVED** [review]
See inline comments. cc @drisspg

**Review by malfet — APPROVED** [review]
It would be good to have a more detailed PR description, that mentions issue where it is reported(already mentioned), PR that caused it, and why passing `other_type` is wrong in this case.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Please address the comments before landing

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:536]: This test passes currently. I think we should check that we no longer warning.

File: test/test_transformers.py
Comment (mikekgfb) [reply test/test_transformers.py:538]: This is the tdefect converted to a test.  I guess the assert was supposed to prove that the dtypes of the twop masks are the same.  Can be removed in the unit test.

File: test/test_transformers.py
Comment (mikekgfb) [reply test/test_transformers.py:536]: Removed, as per above.

File: torch/nn/functional.py
Comment (mikekgfb) [reply torch/nn/functional.py:5210]: See description.  This is for cross checking types of masks.  After mask1 and mask2 have been cross checked, and mask1 has been converted, we would fail the check if we repeat the check at that point.

File: test/test_transformers.py
Comment (mikekgfb) [reply test/test_transformers.py]: done

**Comment by atalman** [comment]
cc @mikekgfb @ydwu4 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176

**Pull request closed without merging**
